types/dispatcher.d.ts

```diff
@@ -36,6 +37,59 @@ declare class Dispatcher extends EventEmitter {
+
+
+  addListener(eventName: 'connect', callback: (origin: URL, targets: Array<Dispatcher>) => void): this;
+  addListener(eventName: 'disconnect', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  addListener(eventName: 'connectionError', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  addListener(eventName: 'drain', callback: (origin: URL) => void): this;
+
+  removeListener(eventName: 'connect', callback: (origin: URL, targets: Array<Dispatcher>) => void): this;
+  removeListener(eventName: 'disconnect', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  removeListener(eventName: 'connectionError', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  removeListener(eventName: 'drain', callback: (origin: URL) => void): this;
+
+  prependListener(eventName: 'connect', callback: (origin: URL, targets: Array<Dispatcher>) => void): this;
+  prependListener(eventName: 'disconnect', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  prependListener(eventName: 'connectionError', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  prependListener(eventName: 'drain', callback: (origin: URL) => void): this;
+
+  prependOnceListener(eventName: 'connect', callback: (origin: URL, targets: Array<Dispatcher>) => void): this;
+  prependOnceListener(eventName: 'disconnect', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  prependOnceListener(eventName: 'connectionError', callback: (origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void): this;
+  prependOnceListener(eventName: 'drain', callback: (origin: URL) => void): this;
+
+  listeners(eventName: 'connect'): ((origin: URL, targets: Array<Dispatcher>) => void)[]
+  listeners(eventName: 'disconnect'): ((origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void)[];
+  listeners(eventName: 'connectionError'): ((origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void)[];
+  listeners(eventName: 'drain'): ((origin: URL) => void)[];
+
+  rawListeners(eventName: 'connect'): ((origin: URL, targets: Array<Dispatcher>) => void)[]
+  rawListeners(eventName: 'disconnect'): ((origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void)[];
+  rawListeners(eventName: 'connectionError'): ((origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void)[];
+  rawListeners(eventName: 'drain'): ((origin: URL) => void)[];
+
+  emit(eventName: 'connect', origin: URL, targets: Array<Dispatcher>): boolean;
+  emit(eventName: 'disconnect', origin: URL, targets: Array<Dispatcher>, error: UndiciError): boolean;
+  emit(eventName: 'connectionError', origin: URL, targets: Array<Dispatcher>, error: UndiciError): boolean;
+  emit(eventName: 'drain', origin: URL): boolean;
```

Comment:
Can you add the tests for those?

Reply:
Sorry I am really having trouble with `tsd`.

I am much more comfortable writing Typescript than this mix. I am finding that the usage of `tsd` instead of writing the library in typescript is actually creating a lot of bugs in the `.d.ts` files (such as `DispatchHandler.onHeaders`).

Would you guys be open to converting the library to typescript instead?

Reply:
We are ok with Undici being JavaScript, thanks for the offer.

Reply:
@mcollina I am unable to write the tests for these events. I am not familiar enough with `tsd`. How would you like to proceed?

Reply:
check out https://github.com/nodejs/undici/tree/main/test/types and tsd README. If you can write TypeScript, you could learn tsd quickly.